Implement SysIdRoutines for subsystems and shooter subsystem and dynamics

srimanachanta opened the pull request:
- implements drive SysId Routines for drive motors
- Removes old polynomial regression characterization command from Akit template
- remove old trajectory generation and following code. This needs to be implemented using new trajectories from [Choreo](https://github.com/SleipnirGroup/Choreo)
- Adds hardware ids for the indexer and intake subsystems
- Removes support for the 2023 offseason robot (RIP)

TODO:
- [x] Finish implementation of the shooter dynamics models
- [x] Finish 3d visualizer of the shooter
- [x] Remove kickup percent from shooter state

srimanachanta committed:
Update deps

srimanachanta committed:
remove support for offseason swerve

srimanachanta committed:
Add shooter

srimanachanta committed:
rip out old stuff

srimanachanta committed:
add sysid routines

srimanachanta committed:
update deps

srimanachanta committed:
todo

srimanachanta committed:
add fixes from saturday's meeting

absolute encoder offset is tweak

srimanachanta committed:
add FF component to shooter module

srimanachanta committed:
Remove Feedback loop of the kickup wheel

just uses raw percentage now, it isnt needed to use feedback for that, its overkill

srimanachanta commented on src/main/java/frc/robot/subsystems/shooter/ShooterDynamics.java:
```diff
@@ -0,0 +1,115 @@
+   *     the note into the shooter.
+   */
+  public record ShooterState(
+      Rotation2d angle, double shooterVelocityMetersPerSecond, double kickupPercent) {
```
The kickup percentage shouldn't be controlled within the ShooterState class. This class should be reserved for setpoint style control and the kickup state should instead be moved directly into the shooter subsystem with methods to control it.

Bluewaves54 replied:
Right away sir

Bluewaves54 committed:
moved kickup percent out of ShooterState into ShooterBase

Bluewaves54 committed:
Finished shooter dynamics for aiming to speaker

Consolidated validRobotState logic and trajectory math into a single static method that returns an Optional

srimanachanta commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -58,47 +52,61 @@ public RobotContainer() {
+  }
+
+  private void configureTunableParameters() {
+    m_autoChooser.addOption(
+        "DriveDynamicForward", m_drive.characterizeDriveDynamic(SysIdRoutine.Direction.kForward));
+    m_autoChooser.addOption(
+        "DriveDynamicReverse", m_drive.characterizeDriveDynamic(SysIdRoutine.Direction.kReverse));
+    m_autoChooser.addOption(
+        "DriveQuasistaticForward",
+        m_drive.characterizeDriveQuasistatic(SysIdRoutine.Direction.kForward));
+    m_autoChooser.addOption(
+        "DriveQuasistaticReverse",
+        m_drive.characterizeDriveQuasistatic(SysIdRoutine.Direction.kReverse));
+    m_autoChooser.addOption(
+        "ErectorDynamicForward",
+        m_shooter.characterizeErectorDynamic(SysIdRoutine.Direction.kForward));
+    m_autoChooser.addOption(
+        "ErectorDynamicReverse",
+        m_shooter.characterizeErectorDynamic(SysIdRoutine.Direction.kReverse));
+    m_autoChooser.addOption(
+        "ErectorQuasistaticForward",
+        m_shooter.characterizeErectorQuasistatic(SysIdRoutine.Direction.kForward));
+    m_autoChooser.addOption(
+        "ErectorQuasistaticReverse",
+        m_shooter.characterizeErectorQuasistatic(SysIdRoutine.Direction.kReverse));
+    m_autoChooser.addOption(
+        "ShooterDynamicForward",
+        m_shooter.characterizeShooterDynamic(SysIdRoutine.Direction.kForward));
+    m_autoChooser.addOption(
+        "ShooterDynamicReverse",
+        m_shooter.characterizeShooterDynamic(SysIdRoutine.Direction.kReverse));
+    m_autoChooser.addOption(
+        "ShooterQuasistaticForward",
+        m_shooter.characterizeShooterQuasistatic(SysIdRoutine.Direction.kForward));
+    m_autoChooser.addOption(
+        "ShooterQuasistaticReverse",
+        m_shooter.characterizeShooterQuasistatic(SysIdRoutine.Direction.kReverse));
```
This selector is gonna get larger, it might make sense to make a FF Characterization Selector that gets polled when test mode is activated

srimanachanta commented on src/main/java/frc/robot/constants/Constants.java:
```diff
@@ -45,4 +48,17 @@ public static RobotMode getRobotMode() {
     else return RobotBase.isReal() ? RobotMode.REAL : RobotMode.REPLAY;
   }
+
+  public static class Shooter {
```
Not sure where ShooterDynamics constants should live. Maybe only put dimensions that don't change per robot here and move everything else into respective classes

Bluewaves54 replied:
I'd say we create a Field class inside Constants that contain all info about the field to be accessed across the code, including info like game element placement, dimensions, etc.

srimanachanta replied:
Yeah, that's what most teams do, FieldConstants

srimanachanta commented on src/main/java/frc/robot/constants/HardwareIds.java:
```diff
@@ -19,25 +19,27 @@ public static class OFFSEASON_2023 {
-    public static final int kBackLeftTurnId = 0; // TODO
-    public static final int kBackLeftDriveId = 0; // TODO
-    public static final int kBackLeftEncoderId = 0; // TODO
+    public static final int kIntakeWristId = 20;
+    public static final int kIntakeRollersId = 21; // TODO
```
Move all indexer/intake related IDs to that PR

srimanachanta commented on src/main/java/frc/robot/constants/HardwareIds.java:
```diff
@@ -19,25 +19,27 @@ public static class OFFSEASON_2023 {
-    public static final int kBackRightTurnId = 0; // TODO
-    public static final int kBackRightDriveId = 0; // TODO
-    public static final int kBackRightEncoderId = 0; // TODO
+    public static final int kIndexerId = 22;
```
ditto on intake IDs

srimanachanta commented on src/main/java/frc/robot/constants/HardwareIds.java:
```diff
@@ -19,25 +19,27 @@ public static class OFFSEASON_2023 {
+    public static final int kIndexerId = 22;
 
-    public static final int kPigeonId = 0; // TODO
+    public static final int kIntakeWristAbsoluteEncoderPort = 0;
+    public static final int kIntakeWristRelativeEncoderAPort = 1;
+    public static final int kIntakeWristRelativeEncoderBPort = 2;
```
ditto on intake IDs

srimanachanta commented on src/main/java/frc/robot/subsystems/shooter/ErectorIOSparkMax.java:
```diff
@@ -0,0 +1,64 @@
+            HardwareIds.COMP_2024.kErectorAbsoluteEncoderPort,
+            HardwareIds.COMP_2024.kErectorRelativeEncoderAPort,
+            HardwareIds.COMP_2024.kErectorRelativeEncoderBPort,
+            Rotation2d.fromRadians(1.1844636870383878 - (Math.PI / 2.0)));
```
Configuring the magnetic offset this way is proving to be weird, this needs to be reviewed more

srimanachanta commented on src/main/java/frc/robot/subsystems/shooter/ShooterBase.java:
```diff
@@ -0,0 +1,288 @@
+
+  private ShooterState m_setpoint = ShooterState.STARTING_STATE;
+
+  // private final SingleJointedMechanismVisualizer m_setpointVisualizer =
+  //     new SingleJointedMechanismVisualizer(
+  //         "Shooter",
+  //         "Setpoint",
+  //         Constants.Shooter.SHOOTER_LENGTH_METERS,
+  //         Constants.Shooter.PIVOT_POSE,
+  //         ShooterState.STARTING_STATE.angle());
+  // private final SingleJointedMechanismVisualizer m_measuredVisualizer =
+  //     new SingleJointedMechanismVisualizer(
+  //         "Shooter",
+  //         "Measured",
+  //         Constants.Shooter.SHOOTER_LENGTH_METERS,
+  //         Constants.Shooter.PIVOT_POSE,
+  //         ShooterState.STARTING_STATE.angle());
```
Finish implementing this

srimanachanta commented on src/main/java/frc/robot/subsystems/shooter/ShooterBase.java:
```diff
@@ -0,0 +1,288 @@
+        shooterModuleKd.initDefault(0.0); // TODO
+      }
+      case ROBOT_SIMBOT -> {
+        erectorKs.initDefault(0.0); // TODO
+        erectorKg.initDefault(0.0); // TODO
+        erectorKv.initDefault(0.0); // TODO
+        erectorKa.initDefault(0.0); // TODO
+        erectorKp.initDefault(0.0); // TODO
+        erectorKi.initDefault(0.0); // TODO
+        erectorKd.initDefault(0.0); // TODO
+        shooterModuleKs.initDefault(0.0); // TODO
+        shooterModuleKv.initDefault(0.0); // TODO
+        shooterModuleKp.initDefault(0.0); // TODO
+        shooterModuleKi.initDefault(0.0); // TODO
+        shooterModuleKd.initDefault(0.0); // TODO
```
SIM coefficients can all be determined in SIM

srimanachanta commented on src/main/java/frc/robot/util/SingleJointedMechanismVisualizer.java:
finish this

srimanachanta commented on src/main/java/frc/robot/subsystems/shooter/ShooterDynamics.java:
finish TODOs and clean up

srimanachanta commented on src/main/java/frc/robot/util/SingleJointedMechanismVisualizer.java:
```diff
@@ -0,0 +1,63 @@
+
+    m_pivotPose = pivotPose;
+
+    m_mechanism = new Mechanism2d(2, 2, new Color8Bit(14, 17, 23));
```
might make sense to make these parameters configurable because they rely on the size of the mechanism itself

srimanachanta commented on src/main/java/frc/robot/util/SingleJointedMechanismVisualizer.java:
```diff
@@ -0,0 +1,63 @@
+    m_mechanismLigament.setAngle(mechanismAngle.getDegrees());
+    Logger.recordOutput(m_mechanismName + "/Mechanism2d/" + m_logKey, m_mechanism);
+
+    System.out.println(Math.toDegrees(mechanismAngle.getRadians() + Math.PI / 2.0));
```
remove print

srimanachanta commented on src/main/java/frc/robot/util/ThroughBoreEncoder.java:
```diff
@@ -0,0 +1,67 @@
+   * @param absoluteEncoderOffset Offset of the duty cycle encoder. This value is subtracted from
+   *     the true absolute position to yield the result.
+   */
+  public ThroughBoreEncoder(
```
nit: rename to `REVThroughBoreEncoder` as there are many types of through bore encoders

Bluewaves54 committed:
refined validRobotPosition checker for shooting note

added a checker the ensure that the hub opening is wide enough for the note from the robot's shooting position

Bluewaves54 committed:
more cleanup

renamed ThroughBore to REVThroughBore
made Mechanism height and width configurable in visualizer

srimanachanta committed:
bump deps

srimanachanta committed:
add 3d visualizer

srimanachanta committed:
Fix pivot pose bug

srimanachanta committed:
add ascope config to readme

srimanachanta committed:
fix magnetic offset for erector

srimanachanta committed:
Optimize CAN bus usage

No longer reports needless encoder port data

srimanachanta committed:
add erector tuning

srimanachanta committed:
i guess too fast

srimanachanta committed:
Fix follower motor issue

srimanachanta committed:
reset erector coefficents cause of follower issue

srimanachanta committed:
make default state the travel state for the shooter

srimanachanta committed:
Make ShooterState struct serializable

and log it

srimanachanta committed:
move shooter model to repo

srimanachanta committed:
organize crap

srimanachanta committed:
Create FieldDimensions.java

srimanachanta committed:
Formatting fixes
